File: ex31/ex31.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let n = nums.length;
    for (let i = n - 1; i > 0; i--) {
        if (nums[i] > nums[i - 1]) {
            let min_index = i;
            for (let j = i + 1; j < n; ++j) {
                if (nums[j] > nums[i - 1] && nums[j] <= nums[min_index])
                    min_index = j;
            }
            swap(nums, i - 1, min_index);
            reverse(nums, i, n - 1);
            return;
        }
    }
    reverse(nums, 0, n - 1);
    return;
};

function swap(nums, i, j) {
    [nums[i], nums[j]] = [nums[j], nums[i]]
}

function reverse(nums, i, j) {
    while (i < j) {
        swap(nums, i, j)
        i++
        j--
    }
}

File: ex31/ex31.py
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        for i in range(n-1, 0, -1):
            if nums[i] > nums[i-1]:
                min_index = i
                for j in range(i+1, n):
                    # print(j, ',', nums[j])
                    if nums[j] > nums[i-1] and nums[j] < nums[min_index]:
                        min_index = j
                # print(nums[min_index])
                nums[i-1], nums[min_index] = nums[min_index], nums[i-1]
                nums[i:] = sorted(nums[i:])
                return
        nums.reverse()

obj = Solution()
nums = [1,3,2]
obj.nextPermutation(nums)
print(nums)